File: node-samples/src/exercise01/coverCalculator.js
/**
 * Exercise 1: Test generation.
 *
 * Calculate the cover amount for a facility based on amount, risk band,
 * and a percentage cap. Used so participants can generate unit tests
 * (happy path, edge cases, invalid input).
 */
import { fileURLToPath } from "node:url";

/**
 * Work out the cover amount from the facility amount and risk band.
 *
 * @param {number} amountGbp - Facility amount in GBP (must be >= 0).
 * @param {string} riskBand - One of "low", "medium", "high".
 * @param {number} percentageCap - Maximum cover as a percentage (0-100).
 * @returns {number} Cover amount in GBP, rounded to 2 decimal places.
 * @throws {Error} If amount is negative, band is invalid, or cap is out of range.
 */
export function calculateCoverAmount(amountGbp, riskBand, percentageCap) {
  if (amountGbp == null || amountGbp < 0) {
    throw new Error("Amount must be non-negative");
  }

  let pct;
  if (riskBand === "low") {
    pct = 50.0;
  } else if (riskBand === "medium") {
    pct = 75.0;
  } else if (riskBand === "high") {
    pct = 90.0;
  } else {
    throw new Error(`Invalid risk band: ${riskBand}`);
  }

  if (
    percentageCap == null ||
    percentageCap < 0 ||
    percentageCap > 100
  ) {
    throw new Error("Percentage cap must be between 0 and 100");
  }

  const effectivePct = Math.min(pct, percentageCap);
  const cover = amountGbp * (effectivePct / 100.0);
  return Math.round(cover * 100) / 100;
}

/**
 * Example usage.
 */
export function main() {
  const result = calculateCoverAmount(100_000.0, "medium", 80.0);
  console.log(`Cover amount: £${result}`);
}

if (process.argv[1] === fileURLToPath(import.meta.url)) main();

File: python-samples/src/samples/exercise01/cover_calculator.py
"""
Exercise 1: Test generation.

Calculate the cover amount for a facility based on amount, risk band,
and a percentage cap. Used so participants can generate unit tests
(happy path, edge cases, invalid input).
"""


def calculate_cover_amount(amount_gbp, risk_band, percentage_cap):
    """
    Work out the cover amount from the facility amount and risk band.

    Args:
        amount_gbp: Facility amount in GBP (must be >= 0).
        risk_band: One of "low", "medium", "high" (percentage of amount covered).
        percentage_cap: Maximum cover as a percentage (0-100), applied after band.

    Returns:
        Cover amount in GBP, rounded to 2 decimal places.

    Raises:
        ValueError: If amount is negative, band is invalid, or cap is out of range.
    """
    if amount_gbp is None or amount_gbp < 0:
        raise ValueError("Amount must be non-negative")

    if risk_band == "low":
        pct = 50.0
    elif risk_band == "medium":
        pct = 75.0
    elif risk_band == "high":
        pct = 90.0
    else:
        raise ValueError(f"Invalid risk band: {risk_band}")

    if percentage_cap is None or percentage_cap < 0 or percentage_cap > 100:
        raise ValueError("Percentage cap must be between 0 and 100")

    effective_pct = min(pct, percentage_cap)
    cover = amount_gbp * (effective_pct / 100.0)
    return round(cover, 2)


def main():
    """Example usage."""
    result = calculate_cover_amount(100_000.0, "medium", 80.0)
    print(f"Cover amount: £{result}")


if __name__ == "__main__":
    main()
